src/simulator.zig

```diff
@@ -797,12 +890,21 @@ pub const Simulator = struct {
                         simulator.options.replica_crash_stability;
                 },
                 .down => {
-                    if (!chance_f64(
+                    // If we are in liveness mode, we need to make sure that all replicas
+                    // (eventually) make it to the same release.
+                    const restart_upgrade =
+                        simulator.replica_releases[replica.replica] <
+                        simulator.replica_releases_limit and
+                        (simulator.core.isSet(replica.replica) or chance_f64(
                         simulator.random,
-                        simulator.options.replica_restart_probability,
-                    )) {
-                        continue;
-                    }
+                        simulator.options.replica_release_catchup_probability,
+                    ));
```

Comment:
Not the prettiest expression! 

Can't suggest any local improvements, perhaps its time to introuce `tick_crash_up` and `tick_crash_down` functions to save on indentation here

Reply:
Good idea. I also modified `tick_crash_down` so that we recompute `recoverable_count` before restarting a replica, rather than tracking it incrementally.